File: language-music-backend/src/schemas/song.py
from typing import Optional

from pydantic import BaseModel

from src.schemas.tasks import LineReorderingTask, WordSelectionTask


class Song(BaseModel):
    spotify_id: str
    youtube_id: Optional[str] = None
    lyrics: str
    lyrics_url: Optional[str] = None


class SongWithLanguage(Song):
    language: str


class ProcessedSong(SongWithLanguage):
    processed_lyrics: str
    word_selection_tasks: list[WordSelectionTask]
    line_reordering_tasks: list[LineReorderingTask]

File: language-music-frontend/src/models/song.ts
export interface LineReorderingTask {
  task_id: number;
  original_line: string;
  scrambled_line: string[];
}

export interface WordSelectionTask {
  task_id: number;
  target_word: string;
  alternatives: string[];
}

export interface Song {
  spotify_id: string;
  youtube_id: string;
  lyrics: string;
  processed_lyrics: string;

  line_reordering_tasks: LineReorderingTask[];
  word_selection_tasks: WordSelectionTask[];
}
